File: shared/src/logger/sentry.ts
/**
 * Sentry initialization for backend services
 * Call this before creating loggers to enable Sentry error tracking
 */

import * as Sentry from '@sentry/node'
import type winston from 'winston'

interface WinstonLogEntry {
  level: string
  message: string
  error?: Error
  service?: string
  agentId?: string
  userId?: string
  conversationId?: string
  timestamp?: string
  [key: string]: unknown
}

export interface SentryConfig {
  dsn: string
  environment?: string
  release?: string
  tracesSampleRate?: number
}

let isInitialized = false

/**
 * Initialize Sentry for backend services
 * Should be called once at application startup
 */
export function initSentry(config: SentryConfig): void {
  if (isInitialized) {
    return
  }

  // Only initialize Sentry in production
  const isProduction = (config.environment || process.env.NODE_ENV) === 'production'
  if (!isProduction) {
    return // Sentry initialization disabled in development environment
  }

  Sentry.init({
    dsn: config.dsn,
    environment: 'production',
    release: config.release || process.env.APP_VERSION,
    tracesSampleRate: config.tracesSampleRate ?? 0.1,
    integrations: [
      // Automatically instrument Node.js libraries and frameworks
      Sentry.httpIntegration(),
      Sentry.expressIntegration(),
    ],
  })

  isInitialized = true
}

/**
 * Get Sentry transport for Winston
 * Returns null if Sentry is not initialized
 */
export function getSentryTransport(): winston.transport | null {
  if (!isInitialized) {
    return null
  }

  // Create custom Winston transport for Sentry
  const WinstonTransport = require('winston-transport')
  
  class SentryWinstonTransport extends WinstonTransport {
    log(info: WinstonLogEntry, callback: () => void): void {
      setImmediate(() => {
        if (info.level === 'error' && info.message) {
          // Capture error with context
          const error = info.error || new Error(info.message)
          
          Sentry.withScope((scope) => {
            // Add metadata as context
            if (info.service) {
              scope.setTag('service', info.service)
            }
            if (info.agentId) {
              scope.setTag('agentId', info.agentId)
            }
            if (info.userId) {
              scope.setTag('userId', info.userId)
            }
            if (info.conversationId) {
              scope.setTag('conversationId', info.conversationId)
            }
            
            // Add extra data
            const extra: Record<string, unknown> = {}
            Object.keys(info).forEach((key) => {
              if (!['level', 'message', 'error', 'service', 'agentId', 'userId', 'conversationId', 'timestamp'].includes(key)) {
                extra[key] = info[key]
              }
            })
            if (Object.keys(extra).length > 0) {
              scope.setExtras(extra)
            }
            
            Sentry.captureException(error)
          })
        }
      })
      
      callback()
    }
  }
  
  return new SentryWinstonTransport() as winston.transport
}

/**
 * Set user context for Sentry
 */
export function setSentryUser(userId: string, email?: string, username?: string): void {
  Sentry.setUser({
    id: userId,
    email,
    username,
  })
}

/**
 * Set additional context (tags, extra data)
 */
export function setSentryContext(context: {
  tags?: Record<string, string>
  extra?: Record<string, unknown>
}): void {
  if (context.tags) {
    Object.entries(context.tags).forEach(([key, value]) => {
      Sentry.setTag(key, value)
    })
  }

  if (context.extra) {
    Sentry.setExtras(context.extra)
  }
}

/**
 * Clear Sentry context
 */
export function clearSentryContext(): void {
  Sentry.setUser(null)
}

export { Sentry }

File: services/voice-agent/src/utils/sentry.py
"""
Sentry initialization for voice agent service
"""

import os
import logging
from typing import Optional

logger = logging.getLogger(__name__)

_sentry_initialized = False

def init_sentry() -> None:
    """
    Initialize Sentry for error tracking
    Should be called once at application startup
    """
    global _sentry_initialized
    
    if _sentry_initialized:
        return
    
    sentry_dsn = os.getenv('SENTRY_DSN')
    if not sentry_dsn:
        logger.info('SENTRY_DSN not set, skipping Sentry initialization')
        return
    
    # Only initialize Sentry in production
    node_env = os.getenv('NODE_ENV', 'development')
    if node_env != 'production':
        return  # Skip Sentry initialization in development
    
    try:
        import sentry_sdk
        from sentry_sdk.integrations.logging import LoggingIntegration
        from sentry_sdk.integrations.fastapi import FastApiIntegration
        
        sentry_sdk.init(
            dsn=sentry_dsn,
            environment='production',
            release=os.getenv('APP_VERSION'),
            traces_sample_rate=0.1,
            
            integrations=[
                LoggingIntegration(
                    level=logging.INFO,        # Capture info and above as breadcrumbs
                    event_level=logging.ERROR  # Send errors as events
                ),
                FastApiIntegration(),
            ],
            
            # Filter out known non-critical errors
            before_send=lambda event, hint: None if _should_filter_error(event, hint) else event,
        )
        
        _sentry_initialized = True
        logger.info('Sentry initialized successfully')
        
    except ImportError:
        logger.warning('sentry-sdk not installed, skipping Sentry initialization')
    except Exception as e:
        logger.error(f'Failed to initialize Sentry: {e}')


def _should_filter_error(event, hint) -> bool:
    """
    Filter out known non-critical errors
    """
    if 'exc_info' in hint and hint['exc_info']:
        exc_type, exc_value, _ = hint['exc_info']
        
        # Filter out expected exceptions
        if isinstance(exc_value, (KeyboardInterrupt, SystemExit)):
            return True
        
        # Filter out validation errors (these are expected)
        if 'ValidationError' in str(exc_type):
            return True
    
    return False


def set_sentry_user(user_id: str, email: Optional[str] = None, username: Optional[str] = None) -> None:
    """
    Set user context for Sentry
    """
    if not _sentry_initialized:
        return
    
    try:
        import sentry_sdk
        sentry_sdk.set_user({
            'id': user_id,
            'email': email,
            'username': username,
        })
    except Exception as e:
        logger.warning(f'Failed to set Sentry user: {e}')


def set_sentry_context(tags: Optional[dict] = None, extra: Optional[dict] = None) -> None:
    """
    Set additional context (tags, extra data)
    """
    if not _sentry_initialized:
        return
    
    try:
        import sentry_sdk
        with sentry_sdk.push_scope() as scope:
            if tags:
                for key, value in tags.items():
                    scope.set_tag(key, value)
            if extra:
                scope.set_extra(extra)
    except Exception as e:
        logger.warning(f'Failed to set Sentry context: {e}')


def capture_exception(error: Exception, **kwargs) -> None:
    """
    Capture an exception to Sentry
    """
    if not _sentry_initialized:
        return
    
    try:
        import sentry_sdk
        sentry_sdk.capture_exception(error, **kwargs)
    except Exception as e:
        logger.warning(f'Failed to capture exception to Sentry: {e}')

